File: Design-Patterns-Using-JavaScript/Strategy/StrategyPatternSorter.js
// Define Sorting interface
class Sorting {
    sort(numbers) {
        throw new Error("sort method must be implemented");
    }
}

// Define BubbleSort class implementing Sorting
class BubbleSort extends Sorting {
    sort(numbers) {
        // Bubble Sort Algorithm
        console.log("Bubble Sort Algorithm executed.");
        const size = numbers.length;
        for (let i = 0; i < size - 1; i++) {
            for (let j = 0; j < size - i - 1; j++) {
                if (numbers[j] > numbers[j + 1]) {
                    // Swapping
                    const temp = numbers[j];
                    numbers[j] = numbers[j + 1];
                    numbers[j + 1] = temp;
                }
            }
        }
    }
}

// Define SelectionSort class implementing Sorting
class SelectionSort extends Sorting {
    sort(numbers) {
        // Selection Sort Algorithm
        console.log("Selection Sort Algorithm executed.");
        const size = numbers.length;
        for (let i = 0; i < size - 1; i++) {
            let maxIndex = 0;
            for (let j = 1; j < size - i; j++) {
                if (numbers[j] > numbers[maxIndex]) {
                    maxIndex = j;
                }
            }
            const temp = numbers[size - 1 - i];
            numbers[size - 1 - i] = numbers[maxIndex];
            numbers[maxIndex] = temp;
        }
    }
}

// Define StrategyClass
class StrategyClass {
    constructor(algo) {
        this.sorter = algo;
    }

    setSorter(algo) {
        this.sorter = algo;
    }

    sort(a) {
        this.sorter.sort(a);
    }
}

// Client code
const a = [4, 5, 3, 2, 6, 7, 1, 8, 9, 10];
const s = new StrategyClass(new BubbleSort());
s.sort(a);
console.log(a);

const b = [4, 5, 3, 2, 6, 7, 1, 8, 9, 10];
s.setSorter(new SelectionSort());
s.sort(b);
console.log(b);

/*
Bubble Sort Algorithm executed.
[ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ]
Selection Sort Algorithm executed.
[ 1, 2, 3, 4, 5, 6, 7, 8, 9, 10 ]
*/

File: Design-Patterns-Using-Ruby/Strategy/StrategyPatternSorter.rb
# Define Sorting interface
module Sorting
    def sort(numbers)
        raise NotImplementedError, 'Subclasses must implement the sort method'
    end
end

# Define BubbleSort class implementing Sorting
class BubbleSort
    include Sorting

    def sort(numbers)
        # Bubble Sort Algorithm
        puts 'Bubble Sort Algorithm executed.'
        size = numbers.length
        (0...size - 1).each do |i|
            (0...size - i - 1).each do |j|
                if numbers[j] > numbers[j + 1]
                    # Swapping
                    numbers[j], numbers[j + 1] = numbers[j + 1], numbers[j]
                end
            end
        end
    end
end

# Define SelectionSort class implementing Sorting
class SelectionSort
    include Sorting

    def sort(numbers)
        # Selection Sort Algorithm
        puts 'Selection Sort Algorithm executed.'
        size = numbers.length
        (0...size - 1).each do |i|
            max_index = 0
            (1...size - i).each do |j|
                max_index = j if numbers[j] > numbers[max_index]
            end
            numbers[size - 1 - i], numbers[max_index] = numbers[max_index], numbers[size - 1 - i]
        end
    end
end

# Define StrategyClass
class StrategyClass
    attr_accessor :sorter

    def initialize(algo)
        @sorter = algo
    end

    def set_sorter(algo)
        @sorter = algo
    end

    def sort(a)
        @sorter.sort(a)
    end
end

# Client code
a = [4, 5, 3, 2, 6, 7, 1, 8, 9, 10]
s = StrategyClass.new(BubbleSort.new)
s.sort(a)
puts a.to_s

b = [4, 5, 3, 2, 6, 7, 1, 8, 9, 10]
s.set_sorter(SelectionSort.new)
s.sort(b)
puts b.to_s

=begin 
Bubble Sort Algorithm executed.
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
Selection Sort Algorithm executed.
[1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
=end